_posts/2016-07-07-better-email-validation-with-obey-and-mailgun.markdown

```diff
@@ -0,0 +1,70 @@
+---
+layout: post
+title: "Better Email Validation with Obey and Mailgun"
+date: 2016-07-07 10:00:00
+categories: asynchronous validation nodejs
+author: kent.safranski
+cover: /assets/images/covers/books.jpg
+description: Email validation seems simple, but you really should be using a service to accurately validate addresses. Here's how Obey and Mailgun can work together to reliably validate emails.
+comments: true
+---
+
+Validating email is simple, right? Just throw a regex at it? Well it's not as simple as you might think. The RFC Spec (Page 27 of [this monster](https://www.ietf.org/rfc/rfc0822.txt), if you're interested) is clear as mud on what constitutes a valid email.
+
+Basically, with all the special characters and conditions of what can go before and after the `@`, a regex to validate emails accurately would be incomprehensable at best. 
+
+_If you'd like to find out more about why email validation sucks check out [this article](http://www.kbedell.com/2011/03/16/how-to-validate-an-email-address-using-regular-expressions/), for now let's move along._
+
+### Using a Service for Validation
+
+We want to know if an email is valid, but a regex won't work, so our best option is a service. [Mailgun's API for email validation](https://documentation.mailgun.com/api-email-validation.html#email-validation) is a wonderful option. It will check a number of things for us:
+
+* General syntax based on RFC
+* DNS validation
+* Spell check (common email services)
+* Service Provider checking
+
+Not only will it take care of the general syntax (what we _would_ use a hyper-complex regex for) but it will make "best-attempt" tries at checking for common errors and valid domains.
+
+So, you may be asking yourself what this looks like. Typically validation is a synchronous task but we're going to call an API to make this work. This is where [Obey](https://github.com/TechnologyAdvice/obey) makes things simple. The promise-first design allows it to handle async and sync requests exactly the same. But what does this look like?
+
+We start with a simple type definition, using Obey and [request-promise](https://www.npmjs.com/package/request-promise) we'll hit the API and handle the results:
+
+{% highlight javascript %}
```

Comment:
Might make sense to take this opportunity to upgrade our Repo to match Github's Jekyll changes. This would become

![screenshot because I don't know how to escape backticks](http://i.tomfro.st/j2EKCW.png)